content/advent-2017/creating-your-own-bot-experience-with-go-sarah.md

```diff
@@ -0,0 +1,241 @@
++++
+author = ["Oklahomer"]
+title = "Creating Your Own Bot Experience with go-sarah"
+linktitle = "Creating Your Own Bot Experience with go-sarah"
+date = 2017-12-04T00:00:00Z
++++
+
+Chat tools have been good friends of developers.
+On chat tools we share our thoughts, problems, solutions, jokes, and pretty much everything we do as software engineers. And when some kind of tasks can be done in the extension of daily chat conversation, chat becomes even more comfortable yet powerful place to stay.
+That is why I think many developers eager to customize their chat experience with bot frameworks.
+In this article I am going to introduce a new golang-based bot framework, [go-sarah](https://github.com/oklahomer/go-sarah), along with its notable characteristics and components.
+
+# Characteristics
+## Conversational Context
+As previously stated, we prefer to execute tasks in the extension of chat-based conversation.
+To maximize that user experience, go-sarah has an idea of "conversational context," which stores current user state and defines what action to follow.
+
+Think about when, like any other day on chat tool, you are talking with your colleagues about a critical issue reported by a user.
+Everybody agrees this must be fixed before Christmas holidays.
+In the extension of this conversation, on chat tool, you type in something like".todo Fix Sarah's issue #123 by 2017-12-20 12:00:00" to register schedule.
+This is somewhat handy, but user still has to input command and its arguments at once.
+There is a room to improve.
+With go-sarah's "conversational context" feature, you can stash currently provided arguments and still prompt user to provide unfilled argument one at a time just like the image below:
+
+![](/postimages/creating-your-own-bot-experience-with-go-sarah/conoversational_context.png)
+
+This is more "conversational" and user friendly. With this feature, not only bot can let user input arguments step by step, bot can validate partial user input and prompt user to re-input.
+This can also be used in a way such as user and AI have conversation during when user inputs ".ai" command until user inputs something like "quit."
+
+## Live Configuration Update
+Let's say you have a command that receive user input such as ".weather San Antonio, Texas" and respond local weather.
+A problem is that this command internally calls a third party web api with expirable token, and the token must be manually updated every once in a while. Do you update configuration file and reboot bot process?
+
+With proper settings, go-sarah reads configuration file when file is updated and applies change to corresponding command's configuration struct in a thread safe manner. So the bot process does not need to be rebooted just to update configuration value. Detail will be covered later on this post.
```

Comment:
Try changing...
"reads configuration file" to "reads a configuration file"
"when file" to "when that file"
"applies change" to "applies changes"
"Detail" to "Details"